0002-RCBS-Crypto-building-block.md

```diff
@@ -0,0 +1,675 @@
+
+The Dapr encryption scheme is optimized for processing data as a stream. Data is chunked into multiple parts which are encrypted independently. This allows us to return data to callers as a stream, even when decrypting messages, being confident that we are not flushing unverified data to the client.
+
+### Key
+
+Each message is encrypted with a 256-bit symmetric **File Key (FK)** that is randomly generated by Dapr for each new message. The key must be generated as 32 byte of output from a CSPRNG (such as Go's `crypto/rand.Reader`) and must not be reused for other files.
+
+The FK is wrapped using a key stored in a key vault (**Key Encryption Key (KEK)**) by Dapr. The result of the wrapping operation is the **Wrapped File Key (WFK)**.  The algorithm used depends on the type of the KEK as well as the algorithms supported by the component: in order of preference:
+
+- Symmetric keys:
+    - AES-KW ([RFC 3394](https://www.rfc-editor.org/rfc/rfc3394.html)): `AES-KW`
+    - AES-GCM-SIV: `AES-GCM-SIV`  
+      This will be implemented if it can be used with keys stored in key vaults; to be confirmed.
+- RSA keys:
+    - RSA OAEP with SHA-256: `RSA-OAEP-256`
+
+> In the future, we should explore how to add support for elliptic curve cryptography, for example P-256/P-384/P-521 or Curve25519, which requires performing a static ECDH key agreement.
+
+### Ciphertext format
+
+The ciphertext is formatted as:
+
+```text
+header || binary payload
+```
+
+### Header
+
+The **header** is human-readable and contains 3 items, each terminated by a line feed (0x0A) character:
+
+1. Name and version of the encryption scheme used. Currently, this is always `dapr.io/enc/v1`
+2. The manifest, which is a JSON object.
+3. The MAC for the header, base64-encoded
+
+> Base64 encoding follows [RFC 4648 §4](https://datatracker.ietf.org/doc/html/rfc4648#section-4) ("standard" format, with padding included but optional when decoding)
+
+```text
+dapr.io/enc/v1
+{"kw":1,"wfk":"hGYjwDpWEXEymSTFZ95zgX8krElb3Gqyls67R8zJA3k=","cph":1,"np":"Y3J5cHRvIQ=="}
+pBDKLrhAWL7IAvDKBV/v7lmbTG6AEZbf3srUN0Pnn30=
+```
+
+#### Manifest
+
+The second line in the header is the **manifest**, which is a compact JSON object.
+
+Its corresponding Go struct is:
+
+```go
+type Manifest struct {
+	// ID of the wrapping algorithm used.
+	// 0x01 = AES-KW
+	// 0x02 = RSA-OAEP-256
+	KeyWrappingAlgorithm int `json:"kw"`
+	// The Wrapped File Key
+	WFK []byte `json:"wfk"`
+	// ID of the cipher used.
+	// 0x01 = AES-GCM
+	// 0x02 = ChaCha20-Poly1305
+	Cipher int `json:"cph"`
+	// Random sequence of 7 bytes generated by a CSPRNG
+	NoncePrefix []byte `json:"np"`
+}
+```
+
+> Dapr will choose AES-GCM as cipher by default.  
```

Comment:
These should probably be configured via the component's YAML, not? It might also vary per implementation since some components will not have all ciphers.

Reply:
No, this is not per-component because this operation always happens in the Dapr sidecar and never in the key vault, for performance reasons.

We implement a sort of "envelope encryption" scheme. The vault is only used to wrap the file key. The actual encryption/decryption of the data happens with the file key within the Dapr sidecar, so we have full control over the algorithms that are offered, and it's independent from the component.